upload: Enable File Upload for users.

Subhasish-Behera opened the pull request:
<!-- See README for documentation, or ask in #zulip-terminal if unclear -->
### What does this PR do, and why?



### External discussion & connections
<!-- [x] all that apply, specifying topic and adding numbers after # for issues/PRs -->
- [x] Discussed in **#zulip-terminal** in `File Upload #T529 #T1414`
- [x] Fully fixes #529 
- [ ] Partially fixes issue #
- [ ] Builds upon previous unmerged work in PR #
- [ ] Is a follow-up to work in PR #
- [ ] Requires merge of PR #
- [ ] Merge will enable work on #

### How did you test this?
<!-- [x] all that apply -->
- [ ] Manually - Behavioral changes
- [ ] Manually - Visual changes
- [ ] Adapting existing automated tests
- [ ] Adding automated tests for new behavior (or missing tests)
- [ ] Existing automated tests should already cover this (*only a refactor of tested code*)

### Self-review checklist for each commit
- [ ] It is a [minimal coherent idea](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development)
- [ ] It has a commit summary following the [documented style](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development) (title & body)
- [ ] It has a commit summary describing the  motivation and reasoning for the change
- [ ] It individually passes linting and tests
- [ ] It contains test additions for any new behavior
- [ ] It flows clearly from a previous branch commit, and/or prepares for the next commit

### Visual changes    <!-- DELETE SECTION IF NO VISUAL CHANGE -->
<!-- Zulip tips at https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html -->
<!-- For video, try asciinema; after uploading, embed using
[![yourtitle](https://asciinema.org/a/<id>.png)](https://asciinema.org/a/<id>)
-->
<!-- NOTE: Attached videos/images will be clearer from smaller terminal windows -->

mounilKshah commented on zulipterminal/core.py:
```diff
@@ -274,6 +291,11 @@ def show_msg_info(
         )
         self.show_pop_up(msg_info_view, "area:msg")
 
+    def show_file_upload_popup(self) -> None:
+        file_upload_view = FileUploadView(self, "Upload File(Enter the location)")
+        self.show_pop_up(file_upload_view, "area:msg")
+
+        return ""  # Supposed to return the Uri
```
There doesn't seem a need for the return statement at the bottom.

mounilKshah commented on zulipterminal/ui_tools/views.py:
```diff
@@ -1192,6 +1192,11 @@ def __init__(
             # urwid.Pile(msg_box.footer),
         )
 
+    def _handle_file_upload(self, file_location):
+        file_location1 = file_location
+        self.uri = self.model.get_file_upld_uri(file_location1)
+        self.controller.set_uri(self.uri)  # Update the uri in the controller
```
This function can be called when `keypress()` detects `ENTER` and the popup can be closed upon encountering this event.

mounilKshah commented on zulipterminal/ui_tools/views.py:
```diff
@@ -1168,7 +1168,32 @@ def __init__(self, controller: Any, title: str) -> None:
+            # urwid.Pile(msg_box.footer),
+        )
 
+    def keypress(self, size: urwid_Size, key: str) -> str:
+        return super().keypress(size, key)
```
Maybe execute a function/some logic when the user hits `ENTER`

Subhasish-Behera replied:
@mounilKshah  The `FileUploadView` is derived from `PopUpView`. `PopUpView` is supposed to get closed on pressing the same key which opened. So I think it is natural that the user wants to close as soon as he has uploaded the file.
As Neil is suggesting considering using something different [(relevant chat)](https://chat.zulip.org/#narrow/stream/206-zulip-terminal/topic/File.20Upload.20.23T529.20.23T1414/near/1607421) than `PopUpView`, we should first discuss a bit about the Base classes to inherit from and then come back to chose appropriate keys and their actions.

mounilKshah commented on zulipterminal/core.py:
```diff
@@ -109,6 +118,15 @@ def __init__(
+    def uri(self) -> str:
+        return self._uri
+
+    def set_uri(self, value: str) -> None:
+        self._uri = value
+        self.uri_updated_event.set()
```
Instead of simply setting the URI, is it possible for you to add the link generated for the uploaded file and append it to the `write_box`? Maybe that can eliminate the need to wait for the upload API to provide response.

mounilKshah reviewed (commented):
Thank you for working on this @Subhasish-Behera
I ran the code from the PR on my local machine. I was able to view the link generated when a file is uploaded by entering the path in the popup.

Further steps can be manipulating the `write_box` such that the generated link gets attached in the message editor.

Subhasish-Behera committed:
keys: Add hotkey `ctrl + o` for File Upload.

Subhasish-Behera committed:
model: Add function to upload file.

get_file_upld_uri takes a file location as an argument and returns the URI.

Subhasish-Behera committed:
views: Introduce FileUploadView to upload files.

FileUploadView's base class is PopUpView. On keypress of `FILE_UPLOAD`
it passes the value of the edit_text to _handle_file_upload function.
The __init__ have write_box as a parameter which is used to insert the
URI in the write_box.The PopUpView is closed on 'ESC' and the upload is
done via 'ENTER'.

Subhasish-Behera committed:
core/boxes: Logic for initial keypress on `FILE_UPLOAD`

show_file_upload_view in core.py creates an instance of FileUploadView
when called. It is called by `ctrl + o`keypress on WriteBox.

Subhasish-Behera committed:
boxes: Add insert_uri method for WriteBox.

insert_uri method takes the file_name and the uri generated for that
file as arguments. The added text is in the format
[<file_name>](/<uri>). At last the cursor is shited after the end of the
added text.

mounilKshah commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -710,6 +710,14 @@ def autocomplete_emojis(
+        edit_widget = self.contents[self.FOCUS_CONTAINER_MESSAGE][
+            self.FOCUS_MESSAGE_BOX_BODY
+        ]
+        edit_widget.edit_text += f"[{file_name}](/{str(uri)})"
```
Upon sending the message  and opening the file, the URL wasn't able to fetch the file. This may be because of the extra `/` before the `uri`, because the URI returned from the server already has a `/` at the start.

Also, the function already has the type of `uri` as `str`, so there might be no need to typecast it to string.

mounilKshah commented on zulipterminal/model.py:
```diff
@@ -557,8 +557,8 @@ def send_stream_message(self, stream: str, topic: str, content: str) -> bool:
 
-    def get_file_upld_uri(self, file_loctn: str) -> str:
-        with open(file_loctn, "rb") as fp:
+    def get_file_upld_uri(self, file_loc: str) -> str:
+        with open(file_loc, "rb") as fp:
```
I think this can be added to the second commit, where the function has been declared?

mounilKshah commented on zulipterminal/ui_tools/views.py:
```diff
@@ -1169,6 +1169,39 @@ def __init__(self, controller: Any, title: str) -> None:
+
+    def _handle_file_upload(self, file_location: str) -> None:
+        self.uri = self.model.get_file_upld_uri(file_location)
+        file_name = file_location.split("/")[-1]
```
This works for now, but I think in case of Windows, the file path contains [backward slashes](https://learn.microsoft.com/en-us/dotnet/standard/io/file-path-formats), and this approach may not work there.

Subhasish-Behera replied:
Here to get the `file_name` we are using the file location and NOT the `url`. So not sure how `urllib` can be used directly to get the file name.

mounilKshah reviewed (changes requested):
This version of the PR is much closer to the bare bones approach which I had mentioned on the stream earlier. The commit structure looks fine to me. I ran this on my local machine, and it worked fine when the extra `/` was not present in the appended URL.

Also, I think it was a good idea to change the position of the cursor after the URL is appended to the compose box content.

neiljp commented on zulipterminal/core.py:
```diff
@@ -274,6 +276,12 @@ def show_msg_info(
 
+    def show_file_upload_popup(self, write_box: WriteBox) -> None:
+        file_upload_view = FileUploadView(
+            self, write_box, "Upload File(Enter the location)"
```
There's no space before the parenthesis here, but let's keep the title simple and drop those and the text inside. The action should be clear from the internals of the popup?

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -1169,6 +1169,39 @@ def __init__(self, controller: Any, title: str) -> None:
+        self.write_box.insert_uri(file_name, self.uri)
+
+    def keypress(self, size: urwid_Size, key: str) -> str:
+        if is_command_key("FILE_UPLOAD", key):
```
You're overriding the behavior of the keypress here, so you should be able to change the key we use here quite easily?

neiljp reviewed (commented):
@Subhasish-Behera As per Mounil's review, this works with a very minor change to the file prefix :+1:

Other than Mounil's and my inline comments, and ensuring the link is generated correctly, the next steps would be to add some tests for commits, in particular those handling the logic. That should help with thinking about the path entered and what the server expects and provides back to the application.

Subhasish-Behera commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -710,6 +710,14 @@ def autocomplete_emojis(
 
         return emoji_typeahead, emojis
 
+    def insert_uri(self, file_name: str, uri: str) -> None:
```
Makes sense.

mounilKshah reviewed (commented):
Ran this code on my local machine and manually did a quick test. It works as per expectation 👍 

I think it is good to start with writing tests for this PR.

neiljp commented:
I left feedback in the stream, so removing needs-review flag, unless there are specific concerns?

Subhasish-Behera committed:
model/views: Handle Invalid File Location and other possible errors.

If a passed file location is invalid then the get_file_upload_uri
returns None and using that the error and reported.

Subhasish-Behera committed:
test_boxes: test for append_uri_and_filename.

Test cases of different file formats.Firstly a initial edit text is set
which is appended by calling append_uri_and_filename.

Subhasish-Behera committed:
test_popups: Tests for FileUploadView.

Firstly the TestWriteBox is imported as unlike other popups
FileUploadView takes a argument write_box. A fixture write_box
is created which creates a instance of TestWriteBox and
returns its write_box fixture.
There are two cases of keypress test- one for keys which dosent make
the popup exit and the other for key that exits the popup.
There are 4 testcases for _handle_file_upload. The tests are based on
if the return of get_file_upload_uri is None or a valid uri. On that
basis the errors/append_func are called or not is asserted.

Subhasish-Behera committed:
test_model: Tests for get_file_upload_uri.

There are three tests based on upload result and existence of file.
tempfile is used to create and write file in case of existing files and
in csase of non-existing files a invalid temp_file_path is created.

mounilKshah commented on tests/model/test_model.py:
```diff
@@ -810,6 +812,49 @@ def test_send_stream_message(
+        self, mocker, model, file_name, upload_result, expected_result
+    ):
+        self.client.upload_file = mocker.Mock(return_value=upload_result)
+        with tempfile.TemporaryDirectory() as temp_dir:
```
Good thinking to use this.

mounilKshah reviewed (commented):
Thank you for implementing all the feedback and writing the tests. I tested the code on my local machine and it worked fine. I even tried uploading multiple files in the same message and that too worked without a glitch.
The tests seem fine to me. I think in the next iteration, apart from the minor suggestion, the commit structure can have the code block commits squashed together with their respective tests.

neiljp commented on tests/model/test_model.py:
```diff
@@ -810,6 +812,49 @@ def test_send_stream_message(
             )
 
+    @pytest.mark.parametrize(
+        "file_name, upload_result, expected_result",
+        [
+            case(
+                "existing_file.txt",
```
I don't see that we gain from using different `file_name` strings, when this variable is passed in as it is each time. Instead I'd suggest using a fixed string - if we continue to set a filename to create - and then using an alternative parameter here which may have a different path. That would allow using the same branch of the conditional each time, and using this new parameter to select the file that is generated or not.

Subhasish-Behera replied:
I probably thought along the lines of testing the code for different file types like `pdf` and `txt`.

Subhasish-Behera replied:
I think you are suggesting that `file_name` doesn't help in testing that.I will change as you said.
 But should ZT-Client's code test if different file formats work?

neiljp replied:
I agree that this illustrates that various file names should be acceptable to the function. However, how many would we include to show that? Also, file extensions do not necessarily correspond to file types - and I suspect the client/server accepts arbitrary files anyhow? So, if we want to demonstrate that, we could include a fixed string in the test - or as a default value to a test parameter - which looks fairly random.

In case it needs clarifying, using a mock for `client.upload` means that only ZT code sees this file in any case - it doesn't touch the actual `Zulip.Client`, and of course not any server.

I mentioned using a path parameter combined with a fixed file_name, since you're not testing eg. `./file_name`, `path_to/file_name` or similar vs just `file_name`. The `file_name` could be a constant, with the path varying in the test cases.

neiljp reviewed (commented):
@Subhasish-Behera Thanks for the updates and the tests :+1:

This does work well, so I've focused more on the details here.

I agree with Mounil that squashing the appropriate test commits with the matching code would make this clearer, and likely also the two commits which build slightly on each other to wrap the upload.

Once your commits are squashed, I'd recommend checking their titles and any summary text, since many have not been updated while you've been working on the PR.

I'm looking forward to having this merged! :)

Subhasish-Behera commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -710,6 +710,14 @@ def autocomplete_emojis(
 
         return emoji_typeahead, emojis
 
+    def append_uri_and_filename(self, file_name: str, uri: str) -> None:
```
will change the order of the arguments.

Subhasish-Behera commented on tests/ui_tools/test_popups.py:
```diff
@@ -872,6 +877,90 @@ def test_keypress_exit_popup(
+                file_name, self.file_upload_view.uri
+            )
+        else:
+            self.controller.append_uri_and_filename.assert_not_called()
```
Got it. will remove it.

neiljp replied:
If you do wish to assert on the mocked details, then my point here was more that the object you're asserting on does not exist - and is different to that a few lines above.

Subhasish-Behera commented on tests/ui_tools/test_popups.py:
```diff
@@ -872,6 +877,90 @@ def test_keypress_exit_popup(
+
+        self.controller.model.get_file_upload_uri.assert_called_once_with(file_location)
+        if not expected_error_message:
+            file_name = Path(file_location).name
```
Cool. I think you are suggesting this
```
    def test__handle_file_upload(
        self,
        file_location: str,
        file_name: str,
        expected_uri: str,
        expected_error_message: Optional[str],
    ) -> None:
```
and directly use the `file_name`  parameter.

neiljp replied:
Right, the principle is that ideally a test is as simple as possible - less logic and calculation - and the test cases handle the input and corresponding expected output (`expected_*` parameters).

neiljp commented on zulipterminal/model.py:
```diff
@@ -560,10 +561,16 @@ def send_stream_message(self, stream: str, topic: str, content: str) -> bool:
-            result = self.client.upload_file(fp)
-        return result["uri"]
+    def get_file_upload_uri(self, file_location: str) -> Optional[str]:
+        if os.path.exists(file_location):
+            with open(file_location, "rb") as fp:
```
Rereading this, it would definitely be better to use the exception approach here, and focus error handling for upload as occurring from the model, compared to error handling with accessing the file from the calling function.

It's possible there could be errors related to processing the upload related to the file handling, but that would be part of the upload process.
